File: soap/spec/gen_structs.py
import sys
import glob
from xml.etree import ElementTree as ET
from collections import OrderedDict

def go_name(s: str):
    no_namespace = s.split('}')[-1] if '}' in s else s
    no_namespace = no_namespace.replace('-', '_')
    parts = no_namespace.split('_')
    go_field = ''.join(part.capitalize() for part in parts)
    # Check if the go_field starts with an unallowed golang variable character
    if go_field and not go_field[0].isalpha():
        go_field = '_' + go_field
    return go_field, no_namespace

def parse_element(elem: ET.Element, namespace: str, structs: dict, is_root: bool=False):
    go_field, xml_name = go_name(elem.tag)

    if structs.get(go_field):
        fields = structs[go_field]
    else:
        fields = dict[str, list[tuple[str, str]]]()
        structs[go_field] = fields

    if elem.text is not None and elem.text.strip():
        fields['Value'] = ('*string', 'xml:",chardata"')

    if is_root:
        fields['XMLName'] = ('xml.Name', f'xml:"{namespace}:{xml_name}"')
        fields['XMLNamespace'] = ('*string', f'xml:"xmlns:{namespace},attr,omitempty"')

    # Add attributes for this element
    for _, (name, _) in enumerate(elem.attrib.items()):
        go_field, xml_name = go_name(name)
        fields[go_field] = ('*string', f'xml:"{namespace}:{xml_name},attr,omitempty"')

    # Add child fields
    for child in elem:
        go_field, xml_name = go_name(child.tag)
        fields[go_field] = (f'*{go_field}', f'xml:"{namespace}:{xml_name},omitempty"')
        parse_element(child, namespace, structs, False)

def gen_struct(name, fields):
    lines = []
    lines.append(f'type {name} struct ' + '{')
    for go_field, (go_type, xml_tag) in fields.items():
        lines.append(f'    {go_field} {go_type} `{xml_tag}`')
    lines.append('}')
    return lines

def write_file(filename, package, structs):
    with open(filename, 'w') as f:
        f.write(f'// Generated from {filename}\n')
        f.write(f'package {package}\n\n')
        f.write('import "encoding/xml"\n\n')
        for _, (name, fields) in enumerate(structs.items()):
            for line in gen_struct(name, fields):
                f.write(line + '\n')
            f.write('\n')

def main(xml_path, namespace):
    files = glob.glob(xml_path)
    for file in files:
        tree = ET.parse(file)
        structs = OrderedDict()
        parse_element(tree.getroot(), namespace, structs, True)
        _, package_name = go_name(tree.getroot().tag).lower()
        file_name = f'../model/{package_name}/{package_name}.go'
        write_file(file_name, package_name, structs)

if __name__ == '__main__':
    if len(sys.argv) != 3:
        print("Usage: python generate_go_structs.py <xml_file_glob> <namespace>")
        sys.exit(1)
    main(sys.argv[1], sys.argv[2])

File: soap/spec/gen_structs.go
// gen_structs.go
// Command-line tool to parse an XML file and generate Go structs for all elements.
// Usage: go run gen_structs.go <xml_file> <namespace> <package>

package main

import (
	"encoding/xml"
	"fmt"
	"io/ioutil"
	"os"
	"regexp"
	"strings"
)

type Field struct {
	Name      string
	Type      string
	Tag       string
	Comment   string
	IsPointer bool
}

type StructDef struct {
	Name   string
	Fields []Field
}

func camelCase(s string) string {
	parts := regexp.MustCompile(`[_-]`).Split(s, -1)
	for i, p := range parts {
		if len(p) > 0 {
			parts[i] = strings.ToUpper(p[:1]) + p[1:]
		}
	}
	return strings.Join(parts, "")
}

func parseElement(e *xmlNode, structs map[string]*StructDef, ns, oldNS string) string {
	name := camelCase(e.XMLName.Local)
	if _, exists := structs[name]; exists {
		return name
	}
	fields := []Field{}
	// Attributes
	for _, attr := range e.Attrs {
		attrName := camelCase(attr.Name.Local)
		fields = append(fields, Field{
			Name:      attrName,
			Type:      "*string",
			Tag:       fmt.Sprintf("xml:\"%s,attr\"", attr.Name.Local),
			IsPointer: true,
		})
	}
	// Children
	for _, child := range e.Children {
		childType := "*string"
		if len(child.Children) > 0 || len(child.Attrs) > 0 {
			childType = "*" + camelCase(child.XMLName.Local)
			parseElement(child, structs, ns, oldNS)
		}
		fields = append(fields, Field{
			Name:      camelCase(child.XMLName.Local),
			Type:      childType,
			Tag:       fmt.Sprintf("xml:\"%s,omitempty\"", child.XMLName.Local),
			IsPointer: true,
		})
	}
	structs[name] = &StructDef{Name: name, Fields: fields}
	return name
}

type xmlNode struct {
	XMLName  xml.Name
	Attrs    []xml.Attr
	Children []*xmlNode
}

func buildTree(d *xml.Decoder, se xml.StartElement) (*xmlNode, error) {
	node := &xmlNode{XMLName: se.Name, Attrs: se.Attr}
	for {
		t, err := d.Token()
		if err != nil {
			return nil, err
		}
		switch tok := t.(type) {
		case xml.StartElement:
			child, err := buildTree(d, tok)
			if err != nil {
				return nil, err
			}
			node.Children = append(node.Children, child)
		case xml.EndElement:
			if tok.Name == se.Name {
				return node, nil
			}
		}
	}
}

func printStruct(s *StructDef, isRoot bool, ns, rootTag string) {
	fmt.Printf("type %s struct {\n", s.Name)
	if isRoot {
		fmt.Printf("\tXMLName xml.Name `xml:\"%s:%s\"`\n", ns, rootTag)
		fmt.Printf("\tXMLNamespace *string `xml:\"xmlns,attr\"`\n")
	}
	used := map[string]bool{}
	for _, f := range s.Fields {
		if used[f.Name] {
			continue
		}
		used[f.Name] = true
		fmt.Printf("\t%s %s `%s` %s\n", f.Name, f.Type, f.Tag, f.Comment)
	}
	fmt.Println("}")
}

func main() {
	if len(os.Args) != 4 {
		fmt.Println("Usage: go run gen_structs.go <xml_file> <namespace> <package>")
		os.Exit(1)
	}
	xmlPath, ns, pkg := os.Args[1], os.Args[2], os.Args[3]
	data, err := ioutil.ReadFile(xmlPath)
	if err != nil {
		panic(err)
	}
	d := xml.NewDecoder(strings.NewReader(string(data)))
	var root *xmlNode
	for {
		t, err := d.Token()
		if err != nil {
			break
		}
		if se, ok := t.(xml.StartElement); ok {
			root, err = buildTree(d, se)
			if err != nil {
				panic(err)
			}
			break
		}
	}
	if root == nil {
		panic("No root element found")
	}
	rootTag := root.XMLName.Local
	structs := map[string]*StructDef{}
	parseElement(root, structs, ns, root.XMLName.Space)
	fmt.Printf("package %s\n\nimport (\n\t\"encoding/xml\"\n)\n\n", pkg)
	order := []string{}
	for name := range structs {
		order = append(order, name)
	}
	for i, name := range order {
		isRoot := i == len(order)-1
		printStruct(structs[name], isRoot, ns, rootTag)
		fmt.Println()
	}
}
